File: typescript/packages/bench/src/services/throughputService.ts
/** Completed-order counting. */

import type { Dataset } from "../dataset";
import { AbstractService } from "../support/abstractService";

/** Completed-order counting. */
export class ThroughputService extends AbstractService {
    /** Service name, used as the audit actor. */
    static readonly NAME = "throughput";

    constructor() {
        super(ThroughputService.NAME);
    }

    /** Orders that reached a billable end state. */
    completed(data: Dataset): number {
        return data.completedOrders().length;
    }
}

File: rust/crates/atelier-bench/src/services/throughput_service.rs
//! Completed-order counting.

use crate::dataset::Dataset;
use crate::support::abstract_service::ServiceBase;

/// Completed-order counting.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ThroughputService {
    base: ServiceBase,
}

impl ThroughputService {
    /// Service name, used as the audit actor.
    pub const NAME: &'static str = "throughput";

    pub fn new() -> Self {
        Self {
            base: ServiceBase::new(Self::NAME),
        }
    }

    /// Shared audit trail.
    pub fn base(&self) -> &ServiceBase {
        &self.base
    }

    /// Mutable access so callers can record their own actions.
    pub fn base_mut(&mut self) -> &mut ServiceBase {
        &mut self.base
    }

    /// Orders that reached a billable end state.
    pub fn completed(&self, data: &Dataset) -> usize {
        data.completed_orders().len()
    }
}

impl Default for ThroughputService {
    fn default() -> Self {
        Self::new()
    }
}
